tests/copy.rs

```diff
@@ -0,0 +1,389 @@
+    let destination = tempdir().context("failed to create temporary directory")?;
+
+    create_random_files(source.path(), 3)
+        .await
+        .context("failed to populate temporary directory")?;
+
+    let config = config();
+    let cancel = CancellationToken::new();
+    for url in urls(&test) {
+        // Copy the local directory to the cloud
+        cloud_copy::copy(
+            config.clone(),
+            source.path(),
+            url.clone(),
+            cancel.clone(),
+            None,
+        )
+        .await
+        .context("failed to upload directory")?;
+
+        // Copy the directory from the cloud to a local directory (delete it first in case it exists)
+        fs::remove_dir_all(destination.path()).context("failed to delete destination directory")?;
+        cloud_copy::copy(
+            config.clone(),
+            url.clone(),
+            destination.path(),
+            cancel.clone(),
+            None,
+        )
+        .await
+        .context("failed to download directory")?;
+
+        let source_entries = walk_dir(source.path()).context("failed to walk source directory")?;
+        let destination_entries =
+            walk_dir(destination.path()).context("failed to walk destination directory")?;
+
+        if source_entries.len() != destination_entries.len() {
+            bail!("source directory and download directory have a different number of entries");
+        }
+
+        for (orig, new) in source_entries.into_iter().zip(destination_entries) {
+            if orig != new {
+                bail!(
+                    "original entry `{orig}` does not match the new entry `{new}`",
+                    orig = orig.display(),
+                    new = new.display()
+                );
+            }
+
+            let source = source.path().join(orig);
+            let destination = destination.path().join(new);
+
+            match (source.is_dir(), destination.is_dir()) {
+                (true, true) => continue,
+                (false, false) => {
+                    // Ensure the uploaded file and the downloaded file are the same
+                    if !same_file_content(&source, &destination)
+                        .await
+                        .context("failed to compare files")?
+                    {
+                        bail!(
+                            "contents of upo file `{source}` does not match the contents of downloaded file `{destination}`",
+                            source = source.display(),
+                            destination = destination.display()
+                        );
+                    }
+                }
+                _ => bail!(
+                    "source entry `{source}` does not match the type of download entry `{destination}`",
+                    source = source.display(),
+                    destination = destination.display()
+                ),
+            }
+        }
+    }
+
+    Ok(())
+}
```

Comment:
Can we just add a small test to make sure when the environment variables are not set that the example URLs for testing don't work?

Reply:
Without the environment variables set, we'll be making network requests to the actual services using bogus credentials for a account/bucket that might actually exist (or otherwise we don't have control over). I'm personally of the opinion that's not a neighborly thing to do from a CI.